File: modules/typed-ids-jackson3/src/main/java/org/framefork/typedIds/bigint/json/jackson/ObjectBigIntIdSerializer.java
package org.framefork.typedIds.bigint.json.jackson;

import tools.jackson.core.JsonGenerator;
import tools.jackson.databind.ValueSerializer;
import tools.jackson.databind.SerializationContext;
import org.framefork.typedIds.bigint.ObjectBigIntId;

public class ObjectBigIntIdSerializer extends ValueSerializer<ObjectBigIntId<?>>
{

    @Override
    public void serialize(
        final ObjectBigIntId<?> id,
        final JsonGenerator jsonGenerator,
        final SerializationContext serializationContext
    )
    {
        jsonGenerator.writeNumber(id.toLong());
    }

}

File: modules/typed-ids/src/main/kotlin/org/framefork/typedIds/bigint/json/kotlinxSerialization/ObjectBigIntIdSerializer.kt
package org.framefork.typedIds.bigint.json.kotlinxSerialization

import kotlinx.serialization.KSerializer
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import org.framefork.typedIds.bigint.ObjectBigIntId

open class ObjectBigIntIdSerializer<T : ObjectBigIntId<*>>(
    private val fromLongConstructor: (Long) -> T,
) : KSerializer<T> {

    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("ObjectBigIntId", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: T) {
        encoder.encodeLong(value.toLong());
    }

    override fun deserialize(decoder: Decoder): T {
        val rawLong = decoder.decodeLong()
        return fromLongConstructor.invoke(rawLong)
    }

}
